### PR TITLE
8378431: Move input validation checks to Java for java.lang.StringUTF16 intrinsics

### PR DESCRIPTION
**Pull request by vy**
Improve input validation checks and documentation of the `java.lang.StringUTF16` intrinsics.

### References:

- [JDK-8374582] #29164 — ditto for `java.lang.StringCoding`
- [JDK-8367129] #28832 — ditto for `java.lang.StringLatin1`

[JDK-8367129]: https://bugs.openjdk.org/browse/JDK-8367129
[JDK-8374582]: https://bugs.openjdk.org/browse/JDK-8374582

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8378431](https://bugs.openjdk.org/browse/JDK-8378431): Move input validation checks to Java for java.lang.StringUTF16 intrinsics (**Sub-task** - P4)

### Reviewers
 * [Damon Fenacci](https://openjdk.org/census#dfenacci) (@dafedafe - Committer)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**) Review applies to [0f5f0a34](https://git.openjdk.org/jdk/pull/30139/files/0f5f0a342f877f40ca243ca8abca51be843b7c97)
 * [Raffaello Giulietti](https://openjdk.org/census#rgiulietti) (@rgiulietti - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/30139/head:pull/30139` \
`$ git checkout pull/30139`

Update a local copy of the PR: \
`$ git checkout pull/30139` \
`$ git pull https://git.openjdk.org/jdk.git pull/30139/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 30139`

View PR using the GUI difftool: \
`$ git pr show -t 30139`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/30139.diff">https://git.openjdk.org/jdk/pull/30139.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/30139#issuecomment-4022734519)
</details>

### CODE REVIEW

File: src/java.base/share/classes/java/lang/StringUTF16.java [hunk 1]
```diff
@@ -67,30 +67,61 @@ static byte[] newBytesFor(int len) {
...
+     * @param val a UTF-16 string byte array
+     * @param index the index of the character to insert the code point to
+     * @param c a code point
+     */
+    // vmIntrinsics::_putCharStringU
     @IntrinsicCandidate
-    // intrinsic performs no bounds checks
     static void putChar(byte[] val, int index, int c) {
-        assert index >= 0 && index < length(val) : "Trusted caller missed bounds check";
+        assert val != null && index >= 0 && index < length(val) : "Trusted caller violated input constraints";
```
Comment (vy) [inline src/java.base/share/classes/java/lang/StringUTF16.java:106]: `putChar`/`getChar` is on several hot spots. Refrained from guarding it further due to performance considerations.

File: test/hotspot/jtreg/compiler/intrinsics/string/TestStringIntrinsicRangeChecks.java [hunk 2]
```diff
@@ -89,20 +89,24 @@ public static void main(String[] args) throws Exception {
         for (int srcOff = 0; srcOff < SIZE; ++srcOff) {
             for (int dstOff = 0; dstOff < SIZE; ++dstOff) {
                 for (int len = 0; len < SIZE; ++len) {
```
Comment (vy) [inline test/hotspot/jtreg/compiler/intrinsics/string/TestStringIntrinsicRangeChecks.java:91]: I'd have preferred dedicated `@Test` methods fuzzing arguments of associated intrinsics using edge cases. For instance, negative (i.e., `< 0`) and excessive (i.e., `> SIZE`) index values are missing in these tests. I avoided these and similar test improvements to keep the PR focused. Maybe in a follow-up PR.

File: src/hotspot/share/opto/library_call.cpp [hunk 3]
```diff
@@ -843,6 +843,24 @@ void LibraryCallKit::set_result(RegionNode* region, PhiNode* value) {
...
+    if (region->req() > 1) { \
+      region = _gvn.transform(region)->as_Region(); \
+      Node* frame = _gvn.transform(new ParmNode(C->start(), TypeFunc::FramePtr)); \
+      Node* halt = _gvn.transform(new HaltNode(region, frame, "unexpected guard failure in intrinsic")); \
+      C->root()->add_req(halt); \
+    } \
+  } while (0); \
+  if (stopped()) { \
+    return true; \
+  }
```
Comment (dafedafe) [inline src/hotspot/share/opto/library_call.cpp]: Since you use the `while(0)` idiom I think it would make sense to move this inside as well.

File: src/hotspot/share/opto/library_call.cpp [hunk 4]
```diff
@@ -843,6 +843,24 @@ void LibraryCallKit::set_result(RegionNode* region, PhiNode* value) {
   assert(value->type()->basic_type() == result()->bottom_type()->basic_type(), "sanity");
 }
 
+#define RETURN_ON_BAILOUT(region, ...) \
```
Comment (dafedafe) [inline src/hotspot/share/opto/library_call.cpp]: I was wondering if it wouldn't be cleaner to avoid creating a macro in this case and instead e.g. have 2 regular methods, one that "initializes the halt nodes" and the other that "finalizes the halt nodes" (it might be slightly better if you need to debug and slightly less risky with variable shadowing... no strong opinion about it though)

**Review by dafedafe — COMMENTED** [review]
Thanks for your continued work on range checks @vy. Overall it look good. I've had a first look and left a couple of inline comments.

File: src/hotspot/share/opto/library_call.cpp (hunk 3 above)
Comment (vy) [reply src/hotspot/share/opto/library_call.cpp]: Changed as requested in e889700699c.

File: src/hotspot/share/opto/library_call.cpp (hunk 4 above)
Comment (vy) [reply src/hotspot/share/opto/library_call.cpp]: > it might be slightly better if you need to debug and slightly less risky with variable shadowing

These are good arguments. I've replaced the `RETURN_ON_BAILOUT` macro with two regular methods in d16b85f8184.

File: src/hotspot/share/opto/library_call.cpp [hunk 5]
```diff
@@ -843,6 +843,22 @@ void LibraryCallKit::set_result(RegionNode* region, PhiNode* value) {
...
+  record_for_igvn(bailout);
+  return bailout;
+}
+
+bool LibraryCallKit::check_bailout(RegionNode* bailout) {
+  if (bailout->req() > 1) {
+    bailout = _gvn.transform(bailout)->as_Region();
+    Node* frame = _gvn.transform(new ParmNode(C->start(), TypeFunc::FramePtr));
+    Node* halt = _gvn.transform(new HaltNode(bailout, frame, "unexpected guard failure in intrinsic"));
+    C->root()->add_req(halt); \
```
Comment (dafedafe) [inline src/hotspot/share/opto/library_call.cpp]:
```suggestion
    C->root()->add_req(halt);
```

File: src/hotspot/share/opto/library_call.cpp [hunk 6]
```diff
@@ -1545,12 +1551,14 @@ bool LibraryCallKit::inline_string_copy(bool compress) {
 #endif //_LP64
 
 //------------------------inline_string_toBytesU--------------------------
-// public static byte[] StringUTF16.toBytes(char[] value, int off, int len)
+// public static byte[] StringUTF16.toBytes0(char[] value, int off, int len)
 bool LibraryCallKit::inline_string_toBytesU() {
   if (too_many_traps(Deoptimization::Reason_intrinsic)) {
     return false;
   }
```
Comment (dafedafe) [inline src/hotspot/share/opto/library_call.cpp]: I think this can be removed as we don't emit an uncommon trap anymore.

File: src/hotspot/share/opto/library_call.cpp
Comment (dafedafe) [inline src/hotspot/share/opto/library_call.cpp]: It can be removed too.

Comment (vy) [inline src/hotspot/share/opto/library_call.cpp]: @dafedafe, can this be removed too?

Comment (vy) [inline src/hotspot/share/opto/library_call.cpp]: @dafedafe, do we still need this too?

Comment (vy) [inline src/hotspot/share/opto/library_call.cpp:1384]: @dafedafe, this one?

Comment (vy) [inline src/hotspot/share/opto/library_call.cpp]: @dafedafe, what about this one?

**Review by RogerRiggs — APPROVED** [review]
Java code changes and tests look fine.  (Did not review VM code).

File: src/hotspot/share/opto/library_call.cpp
Comment (dafedafe) [reply src/hotspot/share/opto/library_call.cpp:1384]: Not this one: there is an `uncommon_trap` below.

Comment (dafedafe) [reply src/hotspot/share/opto/library_call.cpp]: It looks OK.

Comment (dafedafe) [reply src/hotspot/share/opto/library_call.cpp]: This one too.

Comment (dafedafe) [reply src/hotspot/share/opto/library_call.cpp]: This one too.

Note: I made a rough check of the code after the three `too_many_traps` checks that can be removed. It should be OK to remove them but it would be good if you could have another pass as well.

File: src/hotspot/share/opto/library_call.cpp (hunk 5 above)
Comment (vy) [reply src/hotspot/share/opto/library_call.cpp]: Fixed in 0f5f0a342f8.

File: src/hotspot/share/opto/library_call.cpp
Comment (vy) [reply src/hotspot/share/opto/library_call.cpp]: Removed in 0f5f0a342f8.

Comment (vy) [reply src/hotspot/share/opto/library_call.cpp]: Removed in 0f5f0a342f8.

Comment (vy) [reply src/hotspot/share/opto/library_call.cpp:1384]: Right. Not touching it.

Comment (vy) [reply src/hotspot/share/opto/library_call.cpp]: 0f5f0a342f8 implements the recent set of deoptimization-related changes. I've skimmed through it a couple of times, haven't spotted any red flags. I'll report the Tier 1-2 status soon.

File: src/hotspot/share/opto/library_call.cpp (hunk 6 above)
Comment (vy) [reply src/hotspot/share/opto/library_call.cpp]: Removed in 0f5f0a342f8.

File: src/hotspot/share/opto/library_call.cpp
Comment (vy) [reply src/hotspot/share/opto/library_call.cpp]: Removed in 0f5f0a342f8.

File: src/java.base/share/classes/java/lang/StringUTF16.java [hunk 7]
```diff
@@ -67,30 +67,61 @@ static byte[] newBytesFor(int len) {
...
-                                       ", should be less than " + MAX_LENGTH);
+        if (length >= MAX_LENGTH) {
+            throw new OutOfMemoryError("UTF16 String size is " + length +
+                    ", should be less than " + MAX_LENGTH);
         }
-        return len << 1;
     }
 
+    /**
+     * Inserts the given code point to the specified position of the provided
```
Comment (rgiulietti) [inline src/java.base/share/classes/java/lang/StringUTF16.java]: Is there maybe a better term than "inserts"? The latter suggests that all following characters are shifted to make room for the new one.

File: src/java.base/share/classes/java/lang/StringUTF16.java
Comment (rgiulietti) [inline src/java.base/share/classes/java/lang/StringUTF16.java:2043]: Probably not part of this PR, but
```suggestion
    private static void checkBoundsBeginEnd(int begin, int end, byte[] val) {
```

**Review by rgiulietti — COMMENTED** [review]
Otherwise the Java code LGTM.

**Review by RogerRiggs — APPROVED** [review]
Java... still ok.

File: src/java.base/share/classes/java/lang/StringUTF16.java
Comment (vy) [reply src/java.base/share/classes/java/lang/StringUTF16.java:2043]: It cannot be made `private`, it is used by `test/hotspot/jtreg/compiler/patches/java.base/java/lang/Helper.java`.

File: src/java.base/share/classes/java/lang/StringUTF16.java (hunk 7 above)
Comment (vy) [reply src/java.base/share/classes/java/lang/StringUTF16.java]: Right. Replaced it with `writes` in 1f954831abb.

**Review by rgiulietti — APPROVED** [review]
@dafedafe Can you (informally) approve the HotSpot changes? The Java code looks good to both @RogerRiggs and myself.

File: src/hotspot/share/opto/library_call.cpp [hunk 8]
```diff
@@ -1545,12 +1541,10 @@ bool LibraryCallKit::inline_string_copy(bool compress) {
...
 
 //------------------------inline_string_toBytesU--------------------------
-// public static byte[] StringUTF16.toBytes(char[] value, int off, int len)
+// public static byte[] StringUTF16.toBytes0(char[] value, int off, int len)
 bool LibraryCallKit::inline_string_toBytesU() {
-  if (too_many_traps(Deoptimization::Reason_intrinsic)) {
-    return false;
-  }
   // Get the arguments.
+  assert(callee()->signature()->size() == 3, "character array encoder requires 3 arguments");
```
Comment (dafedafe) [inline src/hotspot/share/opto/library_call.cpp:1547]: Good idea! Could we add it to `getChars0` for consistency?

Comment (vy) [reply src/hotspot/share/opto/library_call.cpp:1547]: Good call. 6124a888a4d adds checks for `callee()->signature()->size()` to all `StringUTF16` intrinsics. That is, `getChars0`, `getChar`, and `putChar` is guarded too.

**Review by dafedafe — APPROVED** [review]
LGTM. Thank you @vy.

**Review by rgiulietti — APPROVED** [review]
Thanks @vy
